sklearn/metrics/pairwise.py

```diff
@@ -1138,6 +1111,118 @@ def _pairwise_callable(X, Y, metric, **kwds):
+    or sparse matrices.
+
+    Parameters
+    ----------
+    X : array [n_samples_a, n_samples_a] if metric == "precomputed", or,
+        [n_samples_a, n_features] otherwise
+        Array of pairwise distances between samples, or a feature array.
+
+    Y : array [n_samples_b, n_features], optional
+        An optional second feature array. Only allowed if
+        metric != "precomputed".
+
+    reduce_func : callable, optional
+        The function which is applied on each chunk of the distance matrix,
+        reducing it to needed values.  ``reduce_func(D_chunk, start)``
+        is called repeatedly, where ``D_chunk`` is a contiguous vertical
+        slice of the pairwise distance matrix, starting at row ``start``.
+        It should return an array, a list, or a sparse matrix of length
+        ``D_chunk.shape[0]``, or a tuple of such objects.
+
+        If None, pairwise_distances_chunked returns a generator of vertical
+        chunks of the distance matrix.
+
+    metric : string, or callable
+        The metric to use when calculating distance between instances in a
+        feature array. If metric is a string, it must be one of the options
+        allowed by scipy.spatial.distance.pdist for its metric parameter, or
+        a metric listed in pairwise.PAIRWISE_DISTANCE_FUNCTIONS.
+        If metric is "precomputed", X is assumed to be a distance matrix.
+        Alternatively, if metric is a callable function, it is called on each
+        pair of instances (rows) and the resulting value recorded. The callable
+        should take two arrays from X as input and return a value indicating
+        the distance between them.
+
+    n_jobs : int
+        The number of jobs to use for the computation. This works by breaking
+        down the pairwise matrix into n_jobs even slices and computing them in
+        parallel.
+
+        If -1 all CPUs are used. If 1 is given, no parallel computing code is
+        used at all, which is useful for debugging. For n_jobs below -1,
+        (n_cpus + 1 + n_jobs) are used. Thus for n_jobs = -2, all CPUs but one
+        are used.
+
+    working_memory : int, optional
+        The sought maximum memory for temporary distance matrix chunks.
+        Defaults to sklearn.get_config()['working_memory'].
+
+    `**kwds` : optional keyword parameters
+        Any further parameters are passed directly to the distance function.
+        If using a scipy.spatial.distance metric, the parameters are still
+        metric dependent. See the scipy docs for usage examples.
+
+    Yields
+    ------
+    D_chunk : array or sparse matrix
+        A contiguous slice of distance matrix, optionally processed by
+        ``reduce_func``.
+    """
+    n_samples_X = _num_samples(X)
+    if metric == 'precomputed':
+        slices = (slice(0, n_samples_X),)
+    else:
+        if Y is None:
+            Y = X
+        chunk_n_rows = get_chunk_n_rows(row_bytes=8 * _num_samples(Y),
```

Comment:
`Y` is not square, so I think `row_bytes_` should be `8*Y.shape[1]` not `.shape[0]`. To be more precise couldn't it be something along the lines of,

```py
def _get_row_bytes(X):
    if scipy.sparse.issparse(X):
       # assuming csr
       total_bytes = X.data.nbytes + X.indices.nbytes
    else:
       total_bytes = X.nbytes
    return total_bytes / _num_samples(Y)
```
though this would fail for other input type than dense or csr.

Reply:
I think you've misunderstood.

Each row of a dense (which it should be) distance matrix consists of a float for each row vector in Y.  So the memory we need to account for is len(Y) * 8 bytes per row of X_chunk.

Reply:
The number of features in Y (== in X) may indeed affect temporary memory use, and we could indeed set row_bytes to `8 * (_num_samples(Y) + _num_features(Y))` making some assumptions about how a metric may be implemented. In practice, euclidean has a temporary cost of `X_chunk.shape[0] + Y.shape[0]` for norms; manhattan via scipy does no operations in-place, and so costs `X_chunk.shape[0] * Y.shape[1] * 2` in temporary memory costs. I have now commented in px_dx_chunked that this is not factored in.

If changing the config name to `memory_hint` will help, we can do that...